x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/interceptor/DlsFlsLicenseRequestInterceptor.java

```diff
@@ -56,46 +56,52 @@ public void intercept(
-                    final IndicesAccessControl.DlsFlsUsage dlsFlsUsage = indicesAccessControl.getFieldAndDocumentLevelSecurityUsage();
-                    boolean incompatibleLicense = false;
-                    if (dlsFlsUsage.hasFieldLevelSecurity()) {
-                        logger.debug(
-                            () -> format(
-                                "User [%s] has field level security on [%s]",
-                                requestInfo.getAuthentication(),
-                                indicesAccessControl.getIndicesWithFieldLevelSecurity()
-                            )
-                        );
-                        if (false == FIELD_LEVEL_SECURITY_FEATURE.check(frozenLicenseState)) {
-                            incompatibleLicense = true;
+                    if (logger.isDebugEnabled()) {
+                        final IndicesAccessControl.DlsFlsUsage dlsFlsUsage = indicesAccessControl.getFieldAndDocumentLevelSecurityUsage();
+                        if (dlsFlsUsage.hasFieldLevelSecurity()) {
+                            logger.debug(
+                                () -> format(
+                                    "User [%s] has field level security on [%s]",
+                                    requestInfo.getAuthentication(),
+                                    indicesAccessControl.getIndicesWithFieldLevelSecurity()
+                                )
+                            );
+                        }
+                        if (dlsFlsUsage.hasDocumentLevelSecurity()) {
+                            logger.debug(
+                                () -> format(
+                                    "User [%s] has document level security on [%s]",
+                                    requestInfo.getAuthentication(),
+                                    indicesAccessControl.getIndicesWithDocumentLevelSecurity()
+                                )
+                            );
                         }
                     }
-                    if (dlsFlsUsage.hasDocumentLevelSecurity()) {
-                        logger.debug(
-                            () -> format(
-                                "User [%s] has document level security on [%s]",
-                                requestInfo.getAuthentication(),
-                                indicesAccessControl.getIndicesWithDocumentLevelSecurity()
-                            )
-                        );
-                        if (false == DOCUMENT_LEVEL_SECURITY_FEATURE.check(frozenLicenseState)) {
+                    if (false == DOCUMENT_LEVEL_SECURITY_FEATURE.checkWithoutTracking(frozenLicenseState)
+                        || false == FIELD_LEVEL_SECURITY_FEATURE.checkWithoutTracking(frozenLicenseState)) {
+                        boolean incompatibleLicense = false;
+                        IndicesAccessControl.DlsFlsUsage dlsFlsUsage = indicesAccessControl.getFieldAndDocumentLevelSecurityUsage();
```

Comment:
This triggers the access control computatin but does not track the feature usage.
The code around it makes it to not be a problem.
BUT it would be much easier to follow if we knew that feature usage is tracked whenever the access control map is computed.
Only in this case are the IT tests that check the feature tracker also a test for the optimization that avoids the computation on coordinating nodes.

So my suggestion would be to hook the feature tracking inside the `DlsFlsFeatureTrackingIndicesAccessControlWrapper#getAllIndexPermissions` method. What do you think?

Reply:
We discussed on another channel, my suggestion is not feasible.